### DISCUSSION TITLE
Can we use `make_expression` with mutable reference to update multiple Json Object?

### DISCUSSION THREAD

**Discussion by mapleFU** [danielaparker/jsoncons#466] [Q&A] [answered]
Currently, `jsonpath_expression` type has a `reference` template argument. However, `make_expression` doesn't support construct with a mutable reference.

So:
1. Is it design to just replacing built same json path when `json_query` multiple objects?
2. How can we pre-build a jsonpath expression, and use it both to build and replace?

**Comment by danielaparker** [comment]
I've added a `make_update_expression` to master. It can be used to update the JSON document as follows:

```
    auto expr = jsoncons::jsonpath::make_update_expression<json>("$.books[*]");

    auto callback = [](const std::string& /*path*/, json& book)
    {
        if (book.at("category") == "memoir" && !book.contains("price"))
        {
            book.try_emplace("price", 140.0);
        }
    };

    json doc = json::parse(input);

    expr.evaluate_and_update(doc, callback);

    std::cout << pretty_print(doc) << std::endl;
```

It can also be used to query a JSON document with a callback that takes a `const json&` as a parameter:

```
    auto callback2 = [](const std::string& /*path*/, const json& book)
    {
        if (book.at("category") == "memoir" && !book.contains("price"))
        {
        }
    };

    expr.evaluate(doc, callback2);
```

With this way of constructing a `jsonpath_expression`, the document passed by reference to `evaluate` and `evaluate_and_update` must be `non const`.

I'll investigate whether the `jsonpath_expression` returned by the `make_expression` function can be generalized to support both passing a const reference to `evaluate`, and to support `evaluate_and_update` with a non const reference.

**Reply by mapleFU** [reply]
Thanks! Waiting for new release

**Comment by danielaparker** [comment, marked as answer]
This is supported in 0.172.0, see [jsonpath_expression::update](https://github.com/danielaparker/jsoncons/blob/master/doc/ref/jsonpath/jsonpath_expression/update.md)

**Reply by mapleFU** [reply]
Thanks!
